### ISSUE TITLE
`Engine` should have events `BatchLoaded` and `BatchPrepared`

### ISSUE THREAD

**Issue by philip-bl** [pytorch/ignite#528]
My use case is that I want to log inputs or some statistics of theirs which my model receives to make sure that I haven't messed up.

Currently if I use `create_supervised_trainer` and `create_supervised_evaluator`, I can do this by passing `prepare_batch` argument to them which does the logging. But I don't like the fact that this requires me to create tensorboard logger before creating trainer and evaluator. I think it's valuable to be able to first create the trainer and then attach image logging to it.

**Comment by philip-bl** [comment]
Also I think `prepare_batch` argument in `create_supervised_*` should be a function that takes only batch as argument, without arguments `device` and `non_blocking`, because if an engine is created via `create_supervised_*`, its `device` and `non_blocking` properties don't change in the middle of the program.

**Comment by vfdev-5** [comment]
@philip-bl what do you think about that these events `BatchLoaded` and `BatchPrepared` can be custom for your engine ?

I see, more or less, your point about logging the input batch data. This can be done in various manner, for example:
```python

def update(engine, batch):
    x_original, y_original = batch
    x, y = prepare_batch(batch, device=device, non_blocking=non_blocking)
    y_pred = model(x)
    loss = loss_fn(y_pred, y)
    loss.backward()
    optimizer.step()
    return {
        'images': x_original,
        'targets':  y_original,
        'loss': loss
    }

trainer = Engine(update)

@trainer.on(Events.ITERATION_COMPLETED)
def log_input_batch_stats(engine):
    output = engine.state.output
    # compute and log the stats as you wish
```

Tell me if it works for you.

**Comment by philip-bl** [comment]
@vfdev-5 AFAIK, if `update` returns such dict, I will need to pass `output_transform` to all the loss objects and metric objects. I think that's a suboptimal solution.

Sure, I can write my own engine, I guess. But I think logging inputs is a pretty useful and somewhat commonly used piece of functionality, e.g. Karpathy recommends doing that https://karpathy.github.io/2019/04/25/recipe.

Personally, for this purpose I wrote
```
from dataclass import dataclass, field
from typing import Sequence

@dataclass
class FunctionWithEvents:
    """Each callable in before_call will all the arguments func receives.                                                                                                                            
    Return values of these callables will be ignored.                                                                                                                                                
    Each callable in after_call will receive return value of func as the first argument                                                                                                              
    and arguments func receives as other arguments. Return values will be ignored as well.                                                                                                           
                                                                                                                                                                                                     
    Note that if one of the arguments in a consumable iterable, it might be consumed by                                                                                                              
    functions in before_call before it reached func.                                                                                                                                                 
    """
    func: Callable
    before_call: Sequence[Callable] = field(default_factory=list)
    after_call: Sequence[Callable] = field(default_factory=list)

    def __call__(self, *args, **kwargs):
        for f in self.before_call:
            f(*args, **kwargs)
        return_value = self.func(*args, **kwargs)
        for g in self.after_call:
            g(return_value, *args, **kwargs)
        return return_value
```

I create my own `prepare_batch` function (which is, albeit, the same as the standard `prepare_batch` in ignite), decorate it with `FunctionWithEvents`. Then I can add callbacks to it.

**Comment by vfdev-5** [comment]
@philip-bl I see, yes that's correct about `output_transform`. 

However I do not completely understand the purpose of batch prepared ? Function `prepare_batch` is just moving data to GPU etc. 

Actually, there is maybe another more simple way to intercept the batch as it is stored in `state.batch` :
https://github.com/pytorch/ignite/blob/6b3244265a35b1cac77893c18a78ddd6688f5775/ignite/engine/engine.py#L330-L333

So, the log can happen in ITERATION_STARTED.

**Comment by philip-bl** [comment]
Oh, I didn't know the batch is saved there. That makes everything easier, thanks for the info.
